contract/bridge/contracts/VaultReward.sol

```diff
@@ -0,0 +1,192 @@
+    }
+
+    emit ExtendVaultLockPeriod(_vaultId, oldLockPeriod, vault.lockPeriod);
+  }
+
+  function updateVaultAccClaimableRewards(address _vaultId) public {
+    if (block.timestamp <= lockedVaults[_vaultId].lockExpireAt) {
+      // get vault
+      LockedVault storage vault = lockedVaults[_vaultId];
+
+      // store the old accClaimableRewards
+      uint256 oldAccClaimableRewards = vault.accClaimableRewards;
+
+      // update the vault info
+      (uint256 claimableRewards, uint256 rewardClaimAt) = getClaimableRewards(_vaultId);
+      vault.accClaimableRewards = claimableRewards;
+      vault.rewardClaimAt = rewardClaimAt;
+
+      emit UpdateVaultAccClaimableRewards(_vaultId, oldAccClaimableRewards, vault.accClaimableRewards, rewardClaimAt);
+    }
+  }
+
+  function claimRewards(address payable _vaultId) external {
+    require(_vaultId == msg.sender, "Invalid vaultId");
+
+    // claiim rewards
+    (uint256 claimableRewards, uint256 rewardClaimAt) = _claimRewards(_vaultId, _vaultId);
+
+    emit ClaimRewards(_vaultId, claimableRewards, rewardClaimAt);
+  }
+
+  function claimRewardsAndLock(address payable _vaultId) external {
+    require(_vaultId == msg.sender, "Invalid vaultId");
+
+    // claim rewards
+    (uint256 claimableRewards, uint256 rewardClaimAt) = _claimRewards(_vaultId, address(this));
+
+    // lock collateral
+    IVaultRegistry(oneBtc).lockAdditionalCollateralFromVaultReward{value: claimableRewards}(_vaultId);
+
+    emit ClaimRewardsAndLock(_vaultId, claimableRewards, rewardClaimAt);
+  }
+
+  function _claimRewards(address _vaultId, address payable _to) internal returns(uint256, uint256) {
+    // get reward debt
+    (uint256 claimableRewards, uint256 rewardClaimAt) = getClaimableRewards(_vaultId);
+    
+    // update the vault info
+    LockedVault storage vault = lockedVaults[_vaultId];
+    vault.accClaimableRewards = 0;
+    vault.rewardClaimAt = rewardClaimAt;
+
+    // transfer rewards
+    IVaultReserve(vaultReserve).withdrawReward(_to, claimableRewards);
+
+    return (claimableRewards, rewardClaimAt);
+  }
+
+  function getClaimableRewards(address _vaultId) public view vaultExist(_vaultId) returns (uint256 claimableRewards, uint256 rewardClaimAt) {
+    // get vault collateral
+    (,, uint256 collateral,,,,,, uint256 liquidatedCollateral) = IVaultRegistry(oneBtc).getVault(_vaultId);
+    uint256 vaultUsedCollateral = collateral.sub(liquidatedCollateral);
+
+    // get vault
+    LockedVault memory vault = lockedVaults[_vaultId];
+
+    // get APR based on the lock period
+    uint256 lockPeriod = vault.lockPeriod;
+    uint256 vaultAPR;
+    if (lockPeriod == 3) {
+      vaultAPR = 5;
+    } else if (lockPeriod >= 6) {
+      vaultAPR = 10;
+    } else if (lockPeriod >= 12) {
+      vaultAPR = 15;
+    }
+
+    // get the elapsed time
+    uint256 secPerDay = 60 * 60 * 24;
+    uint256 elapsedSecs;
+    if (block.timestamp <= vault.lockExpireAt) {
+      rewardClaimAt = block.timestamp.sub(
```

Comment:
(not sure what the requirement) but do you need to check if block.timestamp - vault.rewardClaimAt >= 2 weeks. Otherwise, you're effectively shifting the rewardClaimAt back as far as +1 sec, and letting them claim anytime.

Reply:
The rewards can be claimed every 14 days. It's one of the requirements.

Reply:
Thanks. Wouldn't it make sense to add some checks on minimum period to collect?

Reply:
It's checked here.
https://github.com/harmony-one/onebtc/blob/vault-incentive-implementation/contract/bridge/contracts/VaultReward.sol#L182

Reply:
Ah ok.  Seems like if you check for period first, you don't need to mod at all.